pkg/kafka/limits/ingest_limits.go

```diff
@@ -0,0 +1,263 @@
+package limits
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+
+	"github.com/grafana/dskit/services"
+
+	"github.com/grafana/loki/v3/pkg/kafka"
+	"github.com/grafana/loki/v3/pkg/kafka/client"
+	"github.com/grafana/loki/v3/pkg/logproto"
+)
+
+// IngestLimits is a service that manages stream metadata limits.
+type IngestLimits struct {
+	services.Service
+
+	cfg    kafka.IngestLimitsConfig
+	logger log.Logger
+	client *kgo.Client
+
+	// Track stream metadata
+	mtx      sync.RWMutex
+	metadata map[string]map[uint64]int64 // tenant -> streamHash -> lastSeenAt
+}
+
+// NewIngestLimits creates a new IngestLimits service. It initializes the metadata map and sets up a Kafka client
+// if Kafka is enabled in the configuration. The client is configured to consume stream metadata
+// from a dedicated topic with the metadata suffix.
+func NewIngestLimits(cfg kafka.Config, logger log.Logger, reg prometheus.Registerer) (*IngestLimits, error) {
+	var err error
+	s := &IngestLimits{
+		cfg:      cfg.IngestLimits,
+		logger:   logger,
+		metadata: make(map[string]map[uint64]int64),
+	}
+
+	if cfg.IngestLimits.Enabled {
```

Comment:
Would it be better to wrap the call to `NewIngestLimits` with `cfg.IngestLimits.Enabled` instead of put it in `NewIngestLimits`? It feels weird to return a partial initialized struct. Is that a pattern we use elsewhere?